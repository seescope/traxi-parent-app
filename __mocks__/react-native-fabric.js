export const Crashlytics = {
  logException: () => {},
  log: () => {},
  recordError: () => {},
};
