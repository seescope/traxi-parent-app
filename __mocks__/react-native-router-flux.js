import React, { PropTypes } from 'react';
import { View } from 'react-native';

export const Actions = {
  createKid: jest.fn(),
  walkthrough: jest.fn(),
  congratulations: jest.fn(),
  thankyou: jest.fn(),
  areYouReady: jest.fn(),
  intro: jest.fn(),
};

export const Scene = (props) => <View {...props} />;

export const Router = ({ children }) => (
  <View>{children}</View>
);

Router.propTypes = {
  children: PropTypes.array.isRequired,
};
