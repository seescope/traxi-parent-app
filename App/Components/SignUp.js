/* eslint global-require: "off" */

import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { Text, Image } from 'react-native';
import { Actions } from 'react-native-router-flux';
import InAppBilling from 'react-native-billing';
import { Experiment, Variant } from 'react-native-ab';

import Button from '../Components/Button';
import Spacing from '../Components/Spacing';
import { WHITE, TRANSPARENT } from '../Constants/Colours';
import loginWithMethod from '../Actions/LoginWithMethod';
import { selectPrice } from '../Actions/Actions';
import { logError } from '../Utils';


export const handleBillingError = error => {
  if (__DEV__) {
    console.log(`Ignoring InAppBilling error in DEV mode: ${error.message}`);
    return InAppBilling.close().then(Actions.selectDevice());
  }

  logError(`InAppBilling error: ${error.message}`);
  return null;
};

export const beginSetup = price =>
  dispatch => {
    if (price === 'free') {
      return dispatch(loginWithMethod('anonymous'))
        .then(() => Actions.selectDevice());
    }

    return dispatch(loginWithMethod('anonymous'))
      .then(() => InAppBilling.open())
      .then(() => InAppBilling.subscribe(price))
      .then(details => {
        console.log('Successful InAppBilling purchase!', details);
        return InAppBilling.close();
      })
      .then(() => Actions.selectDevice())
      .catch(handleBillingError);
  };

const onSelectPrice = price =>
  dispatch =>
    dispatch(selectPrice(price));

const logoStyle = {
  backgroundColor: TRANSPARENT,
  color: WHITE,
  fontSize: 26,
  textAlign: 'center',
  fontWeight: 'bold',
  fontFamily: 'raleway',
  marginBottom: 16,
};

const subHeaderStyle = {
  backgroundColor: TRANSPARENT,
  color: WHITE,
  fontSize: 13,
  textAlign: 'center',
  fontWeight: 'bold',
  fontFamily: 'raleway',
  marginBottom: 24,
};

const containerStyle = {
  flex: 1,
  alignItems: 'center',
  justifyContent: 'center',
};

const SignUpComponent = ({ onPress, onChoice, price }) =>
  <Image
    source={require('../Images/signup_background.png')}
    style={containerStyle}
    resizeMode="cover"
  >
    <Text style={logoStyle}>Keep your children safe.</Text>
    <Experiment
      name="launch-pricing-experiment"
      onChoice={(experimentId, variantId) => onChoice(variantId)}
    >
      <Variant name="free">
        <Text style={subHeaderStyle}>Monitor unlimited devices free of charge.</Text>
      </Variant>
      <Variant name="subscription_one_dollar">
        <Text style={subHeaderStyle}>Monitor unlimited devices for just $0.99 per month.</Text>
      </Variant>
      <Variant name="subscription_five_dollars">
        <Text style={subHeaderStyle}>Monitor unlimited devices for just $4.99 per month.</Text>
      </Variant>
    </Experiment>

    <Spacing />

    <Button primary onPress={() => onPress(price)}>Start Monitoring</Button>
  </Image>;


const SignUp = connect(
  ({ price }) => ({ price }),
  {
    onPress: beginSetup,
    onChoice: onSelectPrice,
  }
)(SignUpComponent);

SignUpComponent.propTypes = {
  price: PropTypes.string,
  onPress: PropTypes.func.isRequired,
  onChoice: PropTypes.func.isRequired,
};

export default SignUp;
