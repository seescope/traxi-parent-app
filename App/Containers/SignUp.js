// @flow

/* eslint-disable react/jsx-indent */

import React from 'react';
import OneSignal from 'react-native-onesignal';
import { Alert, ScrollView, Text, View } from 'react-native';
import { Actions } from 'react-native-router-flux';
import { connect } from 'react-redux';

import { logError, firstName } from '../Utils';
import Button from '../Components/Button';
import HeaderText from '../Components/HeaderText';
import TextInput from '../Components/TextInput';
import { VERY_LIGHT_GREY, GREY } from '../Constants/Colours';
import STYLES from '../Constants/Styles';
import LoadingIndicator from '../Components/LoadingIndicator';
import Spacing from '../Components/Spacing';

import { startedLoading, stoppedLoading } from '../Reducers/Setup/setupActions';
import * as ParentActions from '../Reducers/Parent/parentActions';

import persistParent from '../AsyncActions/PersistParent';
import fetchReports from '../AsyncActions/FetchReports';
import createParentAuthentication
  from '../AsyncActions/CreateParentAuthentication';

import type { ParentState } from '../Reducers/Parent';
import type { KidsState } from '../Reducers/Kids';
import type { SetupState } from '../Reducers/Setup';

type RootState = {
  parentState: ParentState,
  kidsState: KidsState,
  setupState: SetupState
};

type Props = {
  kidName: string,
  email: ?string,
  onNameChanged: () => {},
  onPasswordChanged: () => {},
  onEmailChanged: () => {},
  onPress: () => {},
  loading: boolean
};

type Fields = {
  name: string,
  email: string,
  password: string
};

type Dispatch = () => Promise<any>;

const style = {
  background: {
    backgroundColor: VERY_LIGHT_GREY,
  },
  outerContainer: {
    flex: 1,
    alignItems: 'center',
    paddingTop: 32,
  },
  container: {
    alignItems: 'center',
  },
  labelText: {
    fontFamily: 'Raleway-ExtraBold',
    fontSize: 14,
    color: GREY,
  },
  buttonContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    alignSelf: 'center',
  },
  headerText: {
    color: GREY,
  },
  innerContainer: {
    marginVertical: 16,
    alignItems: 'flex-start',
  },
};

export const validateFields = ({ name, email, password }: Fields): boolean =>
  !!name &&
  !!email &&
  !!password &&
  name.length > 1 &&
  email.length > 1 &&
  password.length > 1;

export const SignUp = (
  {
    onNameChanged,
    onPasswordChanged,
    onEmailChanged,
    kidName,
    email,
    onPress,
    loading,
  }: Props
) => (
  <ScrollView
    style={style.background}
    contentContainerStyle={style.outerContainer}
  >
    <View style={style.container}>
      <HeaderText style={style.headerText}>Last step!</HeaderText>

      <Spacing height={32} />

      <View style={[STYLES.CARD, style.container]} elevation={6}>
        <View style={style.innerContainer}>
          <Text style={style.labelText}>Your email address:</Text>
          <TextInput
            value={email}
            onChangeText={onEmailChanged}
            autoCapitalize="none"
            keyboardType="email-address"
          />

          <Text style={style.labelText}>Choose a password:</Text>
          <TextInput secureTextEntry onChangeText={onPasswordChanged} />
        </View>
      </View>
    </View>
    <View style={style.buttonContainer}>
      {loading
        ? <LoadingIndicator />
        : <Button primary onPress={onPress}>
            See {kidName}'s usage
          </Button>}
    </View>
  </ScrollView>
);

const mapStateToProps = (rootState: RootState): Object => {
  const { parentState, setupState, kidsState } = rootState;
  const { kidUUID, loading } = setupState;

  if (!kidUUID) throw new Error('No kidUUID!');

  const selectedKid = kidsState[kidUUID];

  return {
    loading,
    email: parentState.email,
    kidName: firstName(selectedKid.name),
  };
};

export const mapDispatchToProps = (dispatch: Dispatch): Object => ({
  onNameChanged: (name: string) => dispatch(ParentActions.setName(name)),
  onEmailChanged: (email: string) => dispatch(ParentActions.setEmail(email)),
  onPasswordChanged: (password: string) =>
    dispatch(ParentActions.setPassword(password)),
  onPress: () => {
    dispatch(startedLoading());
    return dispatch(createParentAuthentication())
      .then(() => dispatch(persistParent()))
      .then(() => dispatch(fetchReports()))
      .then(() => dispatch(stoppedLoading()))
      .then(() => {
        OneSignal.registerForPushNotifications();
        Actions.dashboard({ type: 'reset' });
      })
      .catch(e => {
        dispatch(stoppedLoading());
        logError(e);
        Alert.alert(e.message);
      });
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(SignUp);
