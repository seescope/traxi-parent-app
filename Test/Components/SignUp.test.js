import React from 'react';
import renderer from 'react-test-renderer';

import { Actions } from 'react-native-router-flux';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import InAppBilling from 'react-native-billing';
import { Crashlytics } from 'react-native-fabric';

import SignUp, { beginSetup, handleError } from '../../App/Components/SignUp';

const mockStore = configureStore([thunk]);
const testStore = mockStore({ some: 'state', price: 'test-price' });

const SignUpComponent = () => (
  <Provider store={testStore}>
    <SignUp />
  </Provider>
);

it('renders the <SignUp> component', () => {
  const tree = renderer.create(
    <SignUpComponent />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('does beginSetup with a one-dollar transaction', () => {
  const mockDispatch = jest.fn(() => Promise.resolve());
  return beginSetup('one-dollar')(mockDispatch).then(() => {
    expect(InAppBilling.subscribe.mock.calls).toMatchSnapshot();
    expect(mockDispatch.mock.calls).toMatchSnapshot();
    expect(Actions.selectDevice).toHaveBeenCalled();

    InAppBilling.subscribe.mockClear();
    Actions.selectDevice.mockClear();
  });
});

it('does beginSetup with a free transaction', () => {
  const mockDispatch = jest.fn(() => Promise.resolve());
  return beginSetup('free')(mockDispatch).then(() => {
    expect(InAppBilling.subscribe).not.toHaveBeenCalled();
    expect(mockDispatch.mock.calls).toMatchSnapshot();
    expect(Actions.selectDevice).toHaveBeenCalled();
  });
});

it('handles InAppBilling errors correctly', () => {
  __DEV__ = false;
  const testError = new Error('testing');
  handleError(testError);
  expect(Crashlytics.recordError.mock.calls).toMatchSnapshot();
});
