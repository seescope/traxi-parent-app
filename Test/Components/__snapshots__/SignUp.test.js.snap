exports[`test does beginSetup with a free transaction 1`] = `
Array [
  Array [
    [Function],
  ],
]
`;

exports[`test renders the <SignUp> component 1`] = `
<Image
  resizeMode="cover"
  source={1}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "backgroundColor": "transparent",
        "color": "#FFF",
        "fontFamily": "raleway",
        "fontSize": 26,
        "fontWeight": "bold",
        "marginBottom": 16,
        "textAlign": "center",
      }
    }>
    Keep your children safe.
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "backgroundColor": "transparent",
        "color": "#FFF",
        "fontFamily": "raleway",
        "fontSize": 13,
        "fontWeight": "bold",
        "marginBottom": 24,
        "textAlign": "center",
      }
    }>
    Monitor unlimited devices free of charge.
  </Text>
  <View
    style={
      Object {
        "height": 24,
      }
    } />
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "flexDirection": "row",
        "justifyContent": "center",
        "opacity": 1,
      }
    }
    testID={undefined}>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "rgba(75, 73, 73, 1)",
          "fontSize": 28,
          "fontWeight": "200",
        }
      }>
      Start Monitoring
       
    </Text>
    <Image
      source={1} />
  </View>
</Image>
`;
